Feat: Upgrade SDK 47 and upstream x/evm and x/feemarket module from evmos

GNaD13 opened the pull request:
Close #143

GNaD13 committed:
change dependency

jiujiteiro commented on CHANGELOG.md:
```diff
@@ -101,7 +101,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 - (deps) [6bbb25](https://github.com/realiotech/realio-network/commit/6bbb2584e1d855dba77cde49a415fd4dba282cb5) Bump `cosmos sdk` to [`v0.46.7`](https://github.com/realiotech/cosmos-sdk/releases/tag/v0.46.x-realio-alpha-0.6)
-- (deps) [6bbb25](https://github.com/realiotech/realio-network/commit/6bbb2584e1d855dba77cde49a415fd4dba282cb5) Bump `ethermint` to [`v0.21.0-rc2`](https://github.com/evmos/ethermint/releases/tag/v0.21.0-rc1)
+- (deps) [6bbb25](https://github.com/realiotech/realio-network/commit/6bbb2584e1d855dba77cde49a415fd4dba282cb5) Bump `ethermint` to [`v0.21.0-rc2`](https://github.com/evmos/evmos/v18/releases/tag/v0.21.0-rc1)
```
This should go in the unreleased section above.

jiujiteiro commented on third_party/proto/ethermint/evm/v1/query.proto:
```diff
@@ -8,7 +8,7 @@ import "ethermint/evm/v1/evm.proto";
 import "google/protobuf/timestamp.proto";
 
-option go_package = "github.com/evmos/ethermint/x/evm/types";
+option go_package = "github.com/evmos/evmos/v18/x/evm/types";
```
We should double check that the proto files match the evmos ones. For example

https://github.com/evmos/evmos/blob/v18.0.0/proto/ethermint/evm/v1/query.proto

there are some differences. the rpc query url for example. there is no /etheremint in them.

jiujiteiro commented on third_party/proto/ethermint/evm/v1/evm.proto:
```diff
@@ -3,7 +3,7 @@ package ethermint.evm.v1;
 import "gogoproto/gogo.proto";
 
-option go_package = "github.com/evmos/ethermint/x/evm/types";
+option go_package = "github.com/evmos/evmos/v18/x/evm/types";
```
todo - review all the proto files to make sure we have the same one as the upstream evmos v18 repo

jiujiteiro reviewed (changes requested):
left a few comments. Can you also make sure the lint job is passing?

GNaD13 commented:
> left a few comments. Can you also make sure the lint job is passing?

Yeah, this PR is still on a draft. We will do all the linting after done with the code

neitdung committed:
write upgrade for new minimum commission rate

neitdung committed:
fix lint error

neitdung committed:
add upgrade test and fix error when change commission rate

neitdung committed:
fix lint

neitdung committed:
Fix permission file

neitdung committed:
fix spacing issue

neitdung committed:
specific validator commission in test

GNaD13 committed:
proto third party

GNaD13 committed:
go mod

GNaD13 committed:
remove ethermint dependency

GNaD13 committed:
fork sdk

GNaD13 committed:
app.go

neitdung committed:
fix sdk 47 tests and testutil folder

neitdung committed:
fix sdk 47 cmd folder

GNaD13 committed:
ante.go

GNaD13 committed:
ante

GNaD13 committed:
Merge pull request #5 from decentrio/dzung/sdk-47

Fix sdk 47

neitdung committed:
update sdk 47 x folder

GNaD13 committed:
upgrade

GNaD13 committed:
test_helper

GNaD13 committed:
Merge pull request #6 from decentrio/dzung/sdk-47

update sdk 47 x folder

GNaD13 committed:
proto

GNaD13 committed:
remove third party

GNaD13 committed:
buf

GNaD13 committed:
nit

GNaD13 committed:
nit

GNaD13 committed:
nit

GNaD13 committed:
CI

GNaD13 committed:
nit

GNaD13 committed:
chain ID

GNaD13 committed:
fork evmos

GNaD13 committed:
correct check

GNaD13 committed:
ante test

GNaD13 committed:
TestMintedCoinsEachBlock

GNaD13 committed:
lint

GNaD13 committed:
bump go lint versino

GNaD13 committed:
lint

GNaD13 committed:
lint

neitdung committed:
migrate param space for mint module

GNaD13 committed:
Merge pull request #7 from decentrio/param-space

migrate param space for mint module

GNaD13 committed:
lint

neitdung committed:
Merge branch 'feat/sdk-47' of https://github.com/decentrio/realio-network into fix-min-commission

neitdung committed:
fix for sdk 47

GNaD13 committed:
add RegisterGRPCGatewayRoutes

neitdung committed:
fix lint error

neitdung committed:
Merge branch 'feat/sdk-47' of https://github.com/decentrio/realio-network into fix-min-commission

neitdung committed:
fix lint error

GNaD13 committed:
Merge pull request #8 from decentrio/fix-min-commission

!chore: write upgrade handler for new minimum commission rate

neitdung committed:
upgrade handler for sdk 47

GNaD13 committed:
add tm client ibc

GNaD13 committed:
Merge remote-tracking branch 'decent/feat/sdk-47' into feat/sdk-47

GNaD13 committed:
Merge pull request #9 from decentrio/sdk-47-upgrade

feat: sdk 47 upgrade handler

jiujiteiro commented on app/upgrades/v4/constants.go:
```diff
@@ -0,0 +1,7 @@
+package v4
```
It doesnt look like we ever used the `v2` package namespace. the v2 upgrade directory is empty basically. theres also no v3 unless im missing something.

jiujiteiro replied:
@GNaD13 please see this comment

neitdung replied:
I've changed it to `v3`. You can see it at here: [change upgrade name commit](https://github.com/realiotech/realio-network/pull/144/commits/01c4d0fcb25d0d45c1bbcf501b62cdf55fef1b1d)

jiujiteiro replied:
@neitdung it should be v2 :)

jiujiteiro commented on app/upgrades/v4/upgrades.go:
```diff
@@ -0,0 +1,34 @@
+	cdc codec.BinaryCodec,
+) upgradetypes.UpgradeHandler {
+	return func(ctx sdk.Context, _ upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
+		ctx.Logger().Info("Starting upgrade for multi staking...")
```
update log message

GNaD13 replied:
I updated this

jiujiteiro commented on app/upgrades/v4/param_subspace.go:
```diff
@@ -0,0 +1,59 @@
+	for _, subspace := range pk.GetSubspaces() {
+		var keyTable paramstypes.KeyTable
+		switch subspace.Name() {
+		case authtypes.ModuleName:
```
should we have every module here as a case? for example we do not have the assetModule etc.

GNaD13 replied:
This is the reference for upgrade to sdk 46. We only need to add module that use x/params module. So we no need to add x/asset module here because asset module not use x/params

jiujiteiro commented on app/upgrades/v4/param_subspace.go:
```diff
@@ -0,0 +1,59 @@
+	minttypes "github.com/realiotech/realio-network/x/mint/types"
+)
+
+func migrateParamSubspace(ctx sdk.Context, ck consensuskeeper.Keeper, pk paramskeeper.Keeper) {
```
why do we need this?

GNaD13 replied:
for sdk47, we no more use x/params module so we need to migrate data from x/params. Here is the reference upgrade handlers from sdk46 to sdk47 from sdk team https://github.com/cosmos/cosmos-sdk/blob/b68da64d8e38caa0438dbf3b065d864d91e03c4b/simapp/upgrades.go#L30

jiujiteiro commented on app/upgrades/v4/upgrades.go:
```diff
@@ -0,0 +1,34 @@
+		migrateParamSubspace(ctx, ck, pk)
+		fixMinCommisionRate(ctx, &sk)
+
+		if _, err := ibctmmigrations.PruneExpiredConsensusStates(ctx, cdc, clientKeeper); err != nil {
```
what is the purpose of this in the migration?

GNaD13 replied:
This is the updrade from ibc-v6 to ibc-v7. This will remove all expried client to save storage space [ref](https://github.com/cosmos/ibc-go/blob/v7.2.0/docs/migrations/v6-to-v7.md)

jiujiteiro commented on go.mod:
```diff
@@ -1,213 +1,257 @@
 )
 
 replace (
 	// use Realio sdk v0.46.11-realio-4
```
remove/update comment

jiujiteiro commented on go.mod:
```diff
@@ -1,213 +1,257 @@
-	// use cosmos flavored protobufs
-	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
+	// fork evmos
+	github.com/evmos/evmos/v18 => github.com/decentrio/evmos/v18 v18.0.0-20240509044329-a7bc52adb7bd
```
Can we use the evmos dependency directly? Is there a reason for the decentrio fork?

GNaD13 replied:
Evmos v18 is only allow with [Evmos chainID](https://github.com/evmos/evmos/blob/2f9b757e89be8345134d70f805683830b32aa479/x/evm/keeper/keeper.go#L122) (9000 or 9001). So we need to extend to allow Realio ChainID (3300)

jiujiteiro replied:
@GNaD13 thanks. I will ping them about this

jiujiteiro replied:
@GNaD13 ill have an update on this asap

GNaD13 replied:
Hi @jiujiteiro, I've updated evmos dependency upstream from evmos. Could you check the PR pls?

jiujiteiro commented on third_party/proto/cosmos/auth/v1beta1/genesis.proto:
One of the reasons the `third_party/proto` files existed is so that we can generate the swagger files. check out ./scripts/protoc-swagger-gen.sh. do we have a replacement for this?

GNaD13 replied:
@neitdung will check on this

neitdung replied:
Hi @jiujiteiro, I've added new swagger script without `third_party/proto`, you can check the changes here: [swagger script](https://github.com/decentrio/realio-network/pull/11)

jiujiteiro reviewed (changes requested):
Left some comments to discuss

jiujiteiro commented:
In addition to the comments left in the code, Im curious why we have added all of those files in the tests directory. It looks like they were copied over from the now archived ethermint repo.

GNaD13 committed:
add flag get chain id

GNaD13 committed:
add chain id when initialize app

GNaD13 committed:
nit

GNaD13 committed:
update dependency

GNaD13 committed:
add test cmd

neitdung committed:
Fix store to save new commission

neitdung committed:
Merge branch 'feat/sdk-47' of https://github.com/decentrio/realio-network into sdk-47-upgrade

neitdung committed:
fix lint error

neitdung committed:
fix file permission

GNaD13 committed:
Merge pull request #10 from decentrio/sdk-47-upgrade

Fix store to save new commission

GNaD13 committed:
remove unused code

GNaD13 commented:
> In addition to the comments left in the code, Im curious why we have added all of those files in the tests directory. It looks like they were copied over from the now archived ethermint repo.

yeah, so we just import some testing helpers file from ethermin for realio-network testing. I've removed the unused file

jiujiteiro commented:
@GNaD13 the ethermint repo is not maintained/archived as far as i know. If we think we need to import anything from there lets review please. Ideally Id like to import nothing from it, strictly import from evmos directly. You can see some of the testing deps are out of date for example truffle is no longer maintained. All testing tools we should aim to use hardhat on the evm side.

GNaD13 committed:
change to used evmos test utils

GNaD13 commented:
> @GNaD13 the ethermint repo is not maintained/archived as far as i know. If we think we need to import anything from there lets review please. Ideally Id like to import nothing from it, strictly import from evmos directly. You can see some of the testing deps are out of date for example truffle is no longer maintained. All testing tools we should aim to use hardhat on the evm side.

Yeah, I changed to used code from evmos repository

GNaD13 committed:
nit

neitdung committed:
add new script without third_party folder

neitdung committed:
fix conflict in staking module

GNaD13 committed:
Merge pull request #11 from decentrio/swagger-script

Add new swagger script

neitdung committed:
change upgrade name from v4 to v3

GNaD13 committed:
Merge pull request #12 from decentrio/change-upgrade-name

change upgrade name from v4 to v3

jiujiteiro commented:
@GNaD13 @neitdung once we change the upgrade from v3 to v2, can we add "v2" to the module path in the go.mod?

that way, all the imports will have /realiotech/realio-network/v2.....

this is a pattern to follow on every upgrade for versioning.

GNaD13 commented:
> @GNaD13 @neitdung once we change the upgrade from v3 to v2, can we add "v2" to the module path in the go.mod?
> 
> that way, all the imports will have /realiotech/realio-network/v2.....
> 
> this is a pattern to follow on every upgrade for versioning.

yeah, we got it

GNaD13 committed:
v2

GNaD13 committed:
lint

GNaD13 committed:
nit

GNaD13 committed:
update dependency evmos

GNaD13 committed:
nit

GNaD13 committed:
correct logger name

GNaD13 committed:
change back to use cosmos ante

jiujiteiro commented on app/upgrades/v2/upgrades.go:
```diff
@@ -1,20 +1,37 @@
+
+		fixMinCommisionRate(ctx, sk, stakingLegacySubspace)
+		migrateParamSubspace(ctx, ck, pk)
+		// fixMinCommisionRate(ctx, sk)
```
cleanup

jiujiteiro commented on scripts/protoc-swagger-gen.sh:
```diff
@@ -15,12 +15,13 @@ for dir in $proto_dirs; do
 echo "Generate third_party swagger files"
 
-proto_dirs=$(find ./cosmos ./ethermint ./ibc -path -prune -o -name '*.proto' -print0 | xargs -0 -n1 dirname | sort | uniq)
+proto_dirs=$(find ./cosmos ./ethermint ./ibc ./multistaking ./evmos -path -prune -o -name '*.proto' -print0 | xargs -0 -n1 dirname | sort | uniq)
```
NIT- remove .ethermint proto

jiujiteiro reviewed (commented):
small comments. looks ready pending testing

GNaD13 commented:
Some issue with ante handler, we need to waiting for this [PR](https://github.com/evmos/evmos/pull/2619)

GNaD13 committed:
change to auth ante

GNaD13 committed:
add multi staking params

GNaD13 committed:
change ante to multistaking

GNaD13 committed:
nit

GNaD13 committed:
nit

GNaD13 committed:
remove .ethermint proto

GNaD13 committed:
cleanup

GNaD13 committed:
remove upgrade commission cause we already have a separate upgrade for this

GNaD13 committed:
Merge remote-tracking branch 'origin/main' into feat/sdk-47

GNaD13 committed:
lint

GNaD13 committed:
default genesis multistaking

GNaD13 committed:
nit

GNaD13 committed:
Merge remote-tracking branch 'origin/main' into feat/sdk-47
